Pre-training on large-scale T2w healthy/pathology data

naga-karthik opened the pull request:
This PR adds the code to do supervised (and self-supervised) pretraining on T2w MRI data from a collection of public/private, healthy/pathology datasets. 

TODO: 

- [ ] Add code for self-supervised pre-training
- [x] Test single-gpu and multi-gpu training

valosekj committed:
change 'num_workers' to 0 to prevent 'RuntimeError: received 0 items of ancdata'

valosekj committed:
Update input arg description

valosekj committed:
Rerun the notebook

valosekj committed:
Print hyper-parameters into the log file

valosekj committed:
Add '--cuda' input arg

valosekj committed:
track and save epoch time

valosekj committed:
Plot and save input and output validation images to see how the model is learning

valosekj committed:
Add comment

valosekj committed:
Do not plot 'outputs_v2' as it is a hidden representation

valosekj committed:
Include the epoch number as master title

valosekj committed:
Add 'torch.multiprocessing.set_sharing_strategy('file_system')' to solve number of workers issue

Fixes: https://github.com/ivadomed/model-seg-dcm/issues/8

valosekj committed:
Create validation_figures directory if it does not exist

valosekj committed:
Use 3 leading zeros for the epoch number in the figures fname

valosekj committed:
Link issue

valosekj committed:
Add note for 'RandCropByPosNegLabeld'

It seems that the transform randomly selects a foreground point from image, then use it as center crop. This means that it can find the closest voxel that is just outside the SC and use it as the center (hence it includes the SC).
Source: https://github.com/Project-MONAI/MONAI/issues/452#issuecomment-636065502

valosekj committed:
Add 'number_of_holes' arg to specify the number of holes to be used for the RandCoarseDropoutd and RandCoarseShuffled transforms

valosekj committed:
typo

valosekj committed:
batch_size = 8

valosekj committed:
NUM_WORKERS = batch_size

valosekj committed:
number_of_holes=5

number of holes to be used for the RandCoarseDropoutd and RandCoarseShuffled transforms

valosekj committed:
Update transforms for training of the fine-tuned model

From contrast-agnostic SC seg transforms (which heavily deform the image) to simpler transforms more suitable for lesion seg.
Context: https://github.com/ivadomed/model-seg-dcm/pull/10#discussion_r1527186705

Co-authored-by: Naga Karthik Enamundram <53445351+naga-karthik@users.noreply.github.com>

valosekj committed:
update comment, remove unused imports

valosekj committed:
Add notebook with RandCoarseDropoutd transform debug

valosekj committed:
Fix 'dropout_holes=True' and 'dropout_holes=False' comments

From MONAI doc (https://docs.monai.io/en/latest/transforms.html#randcoarsedropoutd):
- if True, dropout the regions of holes and fill value specified by 'fill_value'
- if False, keep the holes and dropout the outside and fill value specified by 'fill_value'

valosekj committed:
remove unused 'max_spatial_size' arg

valosekj committed:
use 'fill_value=0' for 'RandCoarseDropoutd'

valosekj committed:
Plot also RandCoarseDropoutd dropout_holes=False fill_value=0

valosekj committed:
Add note that the batch size is actually doubled (8*2=16), because we are using two augmented samples per 3D Volume

valosekj committed:
Add '--cuda' input arg

valosekj committed:
Remove 'AsDiscrete'

valosekj committed:
Remove 'AsDiscrete'

valosekj committed:
Add 'CUDA_NUM=args.cuda'

valosekj committed:
Add TODO to increase batch_size to 16

valosekj committed:
Use 'roi_size' for 'define_finetune_train_transforms'

valosekj committed:
Use 'label_sc' to crop samples around the SC

- we have to do  it because the pretrained model was trained on samples of (64, 64, 64)

valosekj committed:
batch_size = 8

valosekj committed:
NUM_WORKERS = batch_size

valosekj committed:
Add 'import torch.multiprocessing'

valosekj committed:
Fix shape logging

valosekj committed:
Change 'img_size' to 'ROI_SIZE'

valosekj committed:
'batch["label"]' --> 'batch["label_lesion"]'

valosekj committed:
Plot and save input and output validation images to see how the model is learning

valosekj committed:
Fix ROI_SIZE for sliding_window_inference

valosekj committed:
Crop samples of 64x64x64 also for Validation of the fine-tuned model

valosekj committed:
update docstring

valosekj committed:
log validation samples shapes

valosekj committed:
Save validation figure only if it contains a lesion

valosekj committed:
Plot GT together with image

valosekj committed:
print unique values in the slice to see if it is binary

valosekj committed:
update output fig fname

valosekj committed:
Add 'AsDiscreted' for Training and  Validation of the fine-tuned model

valosekj committed:
threshold val_labels_list and val_outputs_list by 0.5 threshold before computing validation dice and plotting

naga-karthik committed:
add normalized relu normalization

naga-karthik committed:
fix binarization bug

valosekj committed:
remove 'logger.info(np.unique(output.detach().cpu().numpy()))'

valosekj committed:
overlay prediction over input image

valosekj committed:
Fix variable when getting probabilities from logits

valosekj committed:
Add debug lines

valosekj committed:
PEP8

valosekj committed:
Set validation batch_size to 1

valosekj committed:
improve comments

valosekj committed:
fix figure title

valosekj committed:
comment 'AsDiscreted' transforms

valosekj committed:
Make '--pretrained-model' non required to allow training from the scratch

valosekj committed:
Add script to create spine-generic MSD dataset

valosekj committed:
run notebook again

naga-karthik committed:
add loss script containing loss functions

naga-karthik committed:
add lr scheduler

naga-karthik committed:
add init version of transforms

naga-karthik committed:
add script for dataloading

naga-karthik committed:
add utils for defining models

naga-karthik committed:
working init version of model backbones

naga-karthik committed:
untested version of supervised pretraining code

valosekj commented on pretraining/transforms.py:
```diff
@@ -0,0 +1,121 @@
+            transforms.EnsureChannelFirstd(keys=all_keys),
+            transforms.Orientationd(keys=all_keys, axcodes="RPI"),
+            # NOTE: spine interpolation with order=2 is spline, order=1 is linear
+            transforms.Spacingd(keys=all_keys, pixdim=(1.0, 1.0, 1.0), mode=(2, 1, 1)),
```
`mode=(2, 1, 1)` --> `mode=(2, 1)` since we have only two keys:

```suggestion
            transforms.Spacingd(keys=all_keys, pixdim=(1.0, 1.0, 1.0), mode=(2, 1)),
```

naga-karthik replied:
Nice catch! Done in commit https://github.com/ivadomed/model-seg-dcm/pull/11/commits/c26da42ec393b6406a92cc679dc6fb20ac265d5b

(wanted to fix both the issues in a single commit to avoid redundant commits hence didn't accept commit suggestions)

valosekj commented on pretraining/transforms.py:
```diff
@@ -0,0 +1,121 @@
+            transforms.LoadImaged(keys=all_keys, image_only=False),
+            transforms.EnsureChannelFirstd(keys=all_keys),
+            transforms.Orientationd(keys=all_keys, axcodes="RPI"),
+            transforms.Spacingd(keys=all_keys, pixdim=(1.0, 1.0, 1.0), mode=(2, 1, 1)),
```
`mode=(2, 1, 1)` --> `mode=(2, 1)` since we have only two keys:

```suggestion
            transforms.Spacingd(keys=all_keys, pixdim=(1.0, 1.0, 1.0), mode=(2, 1)),
```

naga-karthik replied:
Done in https://github.com/ivadomed/model-seg-dcm/pull/11/commits/c26da42ec393b6406a92cc679dc6fb20ac265d5b

naga-karthik committed:
fix Spacingd mode for pretraining

naga-karthik committed:
add utils file

naga-karthik committed:
fix typo in pretraining transforms

valosekj committed:
save output to a log file

valosekj committed:
Add new input arg "--datalists" allowing to specify JSON datalist(s) for each dataset. If not provided, all datalists in the '--path-data' folder will be used.

valosekj committed:
Add loop across datalists

valosekj committed:
Add docstring to load_data.loader

valosekj committed:
Various minor arg parse fixes

- make "--path-out" required
- rename "--run_dir" to "--run-dir"
- remove "nargs='+'" for "--path-data" --> it is no longer needed since we expect a single folder with one or more datalists
- PEP8 formatting

valosekj committed:
PEP8 formatting

valosekj committed:
Add comment to @torch.no_grad()

valosekj committed:
Add logger.info that config is loaded

valosekj committed:
rename "type" to "mode" to avoid conflict with built-in name

valosekj committed:
PEP8 formatting

valosekj committed:
rename 'type' --> 'mode'

valosekj committed:
swap the order of val_transforms and inference_transforms for easier code readability

valosekj committed:
remove ununsed imports

valosekj committed:
Okay, one more renaming :-D
'mode' --> 'task' to be consistent main_supervised.py

- sorry for previous commits

valosekj committed:
update loader.load_data docstring

valosekj committed:
Fix transforms.transforms typo

valosekj committed:
Fix 'model_utils' import to 'models.model_utils'

valosekj committed:
Fix 'evaluate' keyword arg name

valosekj committed:
Simplify 'config["autoencoderkl"]["seed"]' --> 'config["seed"]'

- because 'config["seed"]' is already used in models/backbone.py

valosekj committed:
Change use_distributed to False to prevent:

"RuntimeError: Default process group has not been initialized, please make sure to call init_process_group."

- no time to debug this further now

valosekj committed:
Use 'model.run_folder' instead of 'run_folder'

valosekj committed:
Comment 'wandb_run'

valosekj committed:
fix import typo

valosekj committed:
Add 'logger.info(f"Using device: {device}")'

valosekj committed:
rename 'x' to 'batch_data' to make code easier to follow

valosekj committed:
remove unused imports

valosekj committed:
Add an example of the config YAML file using 'SmartFormatter'

valosekj committed:
Comment depths, feature_size, and num_heads for the SwinUNETR model to use defaults

naga-karthik committed:
fix merge conflicts

naga-karthik committed:
fix merge conflicts

naga-karthik committed:
remove  arg in EnsureTyped to fix issue with multi-gpu training

naga-karthik committed:
add dice_score()

naga-karthik committed:
uncomment swinunetr params; improve logging info

naga-karthik committed:
remove unetr as model choice (not supported)

naga-karthik committed:
remove local_rank as arg in favor of os.environ

naga-karthik committed:
use sliding_window_inference for validation

naga-karthik committed:
unwrap model if DDP is used

naga-karthik committed:
log only if local_rank is 0

naga-karthik committed:
fix CUDA error by 'spawn'

naga-karthik committed:
move everything to local_rank

naga-karthik committed:
create run folders

naga-karthik committed:
minor code improvements

naga-karthik committed:
add usage example

naga-karthik committed:
add adaptive wing loss for SC seg pre-training

naga-karthik committed:
switch to AdapWingLoss

naga-karthik committed:
add example yaml config fille

valosekj committed:
Remove 'directories' and 'dataset' keys

valosekj committed:
Remove 'spacing' key because resampling to 1.0x1.0x1.0 spacing is hardcoded in pretraining/transforms.py

valosekj committed:
Add 'train_batch_size' and 'val_batch_size' keys

valosekj committed:
Add note for swinunetr that spatial dimensions of input image must be divisible by 32, i.e., we cannot you, for example, 48x48x48

valosekj committed:
Add depths, num_heads, and feature_size for swinunetr to the example config

valosekj committed:
Update optimizer in the example config

valosekj committed:
Remove train_batch_size and val_batch_size -- they are not used. Instead ["opt"]["batch_size"] is used

valosekj committed:
Add missing space to logger.info

naga-karthik committed:
rename pretraining/ to pretraining_and_finetuning/

naga-karthik committed:
rename main_supervised.py to main_supervised_pretraining.py

naga-karthik committed:
Merge branch 'nk/jv_vit_unetr_ssl' of https://github.com/ivadomed/model-seg-dcm into nk/jv_vit_unetr_ssl

valosekj committed:
Plot and save input and output validation images to see how the model is learning

valosekj committed:
Merge branch 'nk/jv_vit_unetr_ssl' of github.com:ivadomed/model-seg-dcm into nk/jv_vit_unetr_ssl

valosekj committed:
Add script to create MSD-style JSON datalist file for BIDS datasets.

valosekj committed:
Add the `whole-spine` dataset

valosekj committed:
Add note for canproco that we do not use sub-cal* subjects right now due to https://github.com/neuropoly/data-management/issues/305

valosekj committed:
Add note for whole-spine that we do not use this dataset right now due to https://github.com/neuropoly/data-management/issues/306

naga-karthik committed:
add function to plot slices

naga-karthik committed:
move plotting to plot_slices() in utils.py

naga-karthik committed:
crop around SC; resample to 0.8 iso; re-use transforms in validation

naga-karthik committed:
modify crop size -- leave 3rd dim. open

valosekj committed:
PEP8 formatting

valosekj committed:
Plot 10 equally spaced slices

valosekj committed:
Plot all validation samples, not only the first one

valosekj committed:
explicitly mention that the script creates datasets for the SUPERVISED pre-training

- in other words, `image` and `label_sc` keys are used and model is learning to segment the SC

valosekj committed:
fix title placing

valosekj committed:
Use alpha=0.5

valosekj committed:
Add slice number below each prediction slice

valosekj committed:
Reduce space between subplots

valosekj committed:
Move slice number down

valosekj committed:
Move master title down

valosekj committed:
Use dpi=200 to save the val figures

valosekj committed:
change crop_pad_size from [64, 64, -1] to [64, 64, 64] to avoid the following ValueError:

ValueError: spatial dimensions [4] of input image (spatial shape: (64,
64,
-1)) must be divisible by 2**5.

naga-karthik committed:
fix bug swinunetr img_size arg -- set it to patch_size

valosekj committed:
PEP8 formatting

valosekj committed:
PEP8 formatting

valosekj committed:
Rename 'InitWeights_He' --> 'InitWeightsHe' to supress warning that a Class should be CapWords

valosekj committed:
Logging refactoring: log everything into a log inside model.run_folder

To do so, I:
- moved model building at the beginning of 'main_worker'
- added 'model.log_dir' variable into the BackboneModel class
- moved logger initialization ('logger.add') into the BackboneModel class

valosekj committed:
PEP8

valosekj committed:
move 'set_determinism(config["seed"])' up

valosekj committed:
Move initializing of the distributed training process before DDP

valosekj committed:
Read log_dir from the model.log_dir before DDP

valosekj committed:
Add 'logger.info("Getting datalists ...")'

valosekj committed:
Plot only a single random validation image each epoch to limit number of validation figures

valosekj committed:
Remove unused imports

valosekj committed:
Fix the logic to get random val sample to plot

valosekj committed:
Remove space

naga-karthik committed:
remove redundant if args.dist statements

valosekj committed:
Add note why we use "if local_rank == 0 else None"

valosekj committed:
Add docstring for 'plot_slices'

naga-karthik committed:
set 3rd dim to -1 when cropping/resizing

naga-karthik committed:
add wip version of finetuning script

naga-karthik committed:
update finetuning transforms

naga-karthik committed:
fix bug in replacing prefixess state_dict's keys when loading pretrained model

naga-karthik committed:
log train/val dice score to the terminal along with loss

naga-karthik committed:
add RandCropByPosNegLabeld transform to finetuning

naga-karthik committed:
log train/val dice score to terminal in finetuning

naga-karthik committed:
improve logging based on finetuning or from scratch

naga-karthik committed:
fix bug in model loading

valosekj commented on pretraining_and_finetuning/transforms.py:
```diff
@@ -0,0 +1,142 @@
+            #             scale_range=(-0.2, 0.2),   
+            #             translate_range=(-0.1, 0.1)),
+            # transforms.Rand3DElasticd(keys=all_keys, prob=0.5, sigma_range=(3.5, 5.5), magnitude_range=(25., 35.)),
+            transforms.RandSimulateLowResolutiond(keys=["image"], zoom_range=(0.5, 1.0), prob=0.25),
+            transforms.RandAdjustContrastd(keys=["image"], gamma=(0.5, 3.), prob=0.25),  # this is monai's RandomGamma
+            # transforms.RandBiasFieldd(keys=["image"], coeff_range=(0.0, 0.5), degree=3, prob=0.3),
+            # transforms.RandGaussianNoised(keys=["image"], mean=0.0, std=0.1, prob=0.1),
+            transforms.RandGaussianSmoothd(keys=["image"], sigma_x=(0., 2.), sigma_y=(0., 2.), sigma_z=(0., 2.0),
+                                           prob=0.3),
+            transforms.RandScaleIntensityd(keys=["image"], factors=(-0.25, 1), prob=0.15),
```
Just to double-check; these four transformations are done only on `image`, right?

valosekj committed:
Add notes

valosekj committed:
remove unused imports

valosekj committed:
improve 'raise ValueError' and 'logger.info' messages

valosekj committed:
remove spaces before '...' during logging

valosekj committed:
Add 'self.task' to the class 'BackboneModel' to track whether we do pretraining or finetuning_pretrained or finetuning_scratch

valosekj committed:
Use 'self.model_name' instead of 'model_name'

valosekj committed:
Use 'self.model_name' instead of 'model_name'

valosekj committed:
Add 'logger.info(f'Output directory: {self.log_dir}')'

valosekj committed:
Remove unused 'device' param from 'train_transforms'

valosekj committed:
Use """ instead of ''' for docstring

naga-karthik committed:
add epochs to run folder name

valosekj committed:
Add 'task' input arg (choices: pretraining or finetuning) to the 'plot_slices' function.

 WHY?
 - For pretraining (which is done on spinal cord segmentations), we plot 10 equally spaced slices because spinal cord usually spans the entire patch, i.e., all slices contain the spinal cord.
 - For finetuning (which is done on lesions), we plot only slices that contain lesions (i.e., non-zero labels). Because the lesion is usually small and covers only a few slices.

valosekj committed:
Add TODO that we may want to plot sagittal slices for sagittal images

valosekj committed:
If the label is empty, exit the function

valosekj committed:
Add logging

valosekj committed:
Add 'num_slices_to_plot = len(slice_indices)'

valosekj committed:
Fix plotting logic

valosekj committed:
Improve logging message

valosekj committed:
Copy the config file to the log directory to keep track of the training details

valosekj committed:
Log the total number of the training and validation samples

valosekj committed:
PEP8

valosekj committed:
PEP8

valosekj committed:
Add TODO to limit multi-processing warnings

valosekj committed:
Plot training and validation loss and dice values

valosekj committed:
Add comments, clarify variable name

valosekj committed:
Use '.detach().cpu()'

valosekj committed:
Make fontsize smaller

valosekj committed:
Plot training and validation loss also for finetuning

valosekj committed:
Make "--resume-from-checkpoint" and "--run-dir" input args properly working

- "--run-dir" was not used at all-- this is fixed now
- "start_epoch" variable was not used at all-- this is fixed now

valosekj committed:
Add logging when the new best model saved

valosekj committed:
Comment 'transforms.CropForegroundd'

Pre-training with 'transforms.CropForegroundd' was crashing to zero. Details: https://github.com/ivadomed/model-seg-dcm/issues/12

valosekj committed:
Change crop_pad_size to [64, 160, 320] to not crop SC

valosekj committed:
Save the latest checkpoint.pth after each epoch

Note that I renamed "best_loss" to "loss" inside the "checkpoint" directory

valosekj committed:
Remove logging validation loss when saving checkpoint.pth

- validation loss only for best_model.pth which is saved if val_loss < best_loss

valosekj committed:
Add sci-zurich

naga-karthik committed:
add training params for monai-unet

naga-karthik committed:
change max_epochs; update batch_size

naga-karthik committed:
uncomment affine transform

naga-karthik committed:
use DiceCELoss instead of AdapWing

naga-karthik committed:
move normalization up before calculating loss

valosekj committed:
Fix mode for transforms.RandAffined

valosekj committed:
uncomment transforms.CropForegroundd

valosekj committed:
Save 'checkpoint.pth' after each epoch

valosekj committed:
Make "--resume-from-checkpoint" and "--run-dir" input args properly working

valosekj committed:
limit the number of slices to plot to 10

valosekj committed:
Change transforms.Spacingd to 1.0 iso

valosekj committed:
comment transforms.CropForegroundd. Again :-D

valosekj committed:
uncomment transforms

valosekj committed:
change crop_pad_size and epoch nb

valosekj committed:
do not hardcode 'config["model"]["monai-unet"]["spatial_dims"]'

valosekj committed:
fix plotting logic

valosekj committed:
Run 'plot_slices' only if local_rank == 0

valosekj committed:
Add 'local_rank' to BackboneModel to do not duplicate logger.info

valosekj committed:
Fix 'local_rank' passing to 'load_data' to do not duplicate logger.info

valosekj committed:
Use 'local_rank == 0' to not duplicate logger.info

valosekj committed:
Copy the transforms.py file to the log directory to keep track of the training/validation transforms

valosekj committed:
Use 'if local_rank == 0 else None' and add 'logger.info'

valosekj committed:
Rotate val samples by 90 degrees before plotting

valosekj committed:
logg loss function

valosekj committed:
Add dropout to 'monai-unet'

valosekj committed:
Add a comment why we use 320 channels

valosekj committed:
Copy datalists to the log directory

valosekj committed:
Fix typo

valosekj committed:
Add 'git annex get $(find . -name "*T2w_lesion-manual.nii.gz")'

valosekj committed:
Rotate by 270 degrees instead of 90

valosekj committed:
Add attentionunet

valosekj committed:
Add attentionunet

valosekj committed:
Fix suptitle position

valosekj committed:
remove num_res_units for AttentionUnet
